File: go/src/deposit/src/get.go
package deposit

import (
	"errors"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
)

func GetDepositByUserId(userId string) ([]Deposit, error) {
	out, err := DdbSvc.Query(&dynamodb.QueryInput{
		TableName: TableName,
		IndexName: aws.String("userIndex"),
		KeyConditions: map[string]*dynamodb.Condition{
			"userId": {
				ComparisonOperator: aws.String("EQ"),
				AttributeValueList: []*dynamodb.AttributeValue{
					{S: aws.String(userId)},
				},
			},
		},
	})
	if err != nil {
		return nil, err
	}
	var deposits []Deposit
	if err := dynamodbattribute.UnmarshalListOfMaps(out.Items, &deposits); err != nil {
		return nil, err
	}
	return deposits, nil
}

func GetDepositById(id string) (Deposit, error) {
	if id == "" {
		return Deposit{}, errors.New("Deposit not found")
	}
	out, err := DdbSvc.GetItem(&dynamodb.GetItemInput{
		TableName:      TableName,
		ConsistentRead: aws.Bool(true),
		Key: map[string]*dynamodb.AttributeValue{
			"depositId": {S: aws.String(id)},
		},
	})
	if err != nil {
		return Deposit{}, err
	}
	if out.Item == nil {
		return Deposit{}, errors.New("Deposit not found")
	}
	var d Deposit
	if err := dynamodbattribute.UnmarshalMap(out.Item, &d); err != nil {
		return Deposit{}, err
	}
	return d, nil
}

File: python/src/deposit/get.py
import os
from typing import Any

from deposit.serde import deposit_from_ddb_item
from deposit.structs import Deposit
from utils.awsutil import new_dynamo_client


def _table_name() -> str:
    name = os.environ.get("TABLE_NAME")
    if not name:
        raise RuntimeError("TABLE_NAME env var is required")
    return name


def _ddb() -> Any:
    return new_dynamo_client()


def get_deposit_by_user_id(user_id: str) -> list[Deposit]:
    response = _ddb().query(
        TableName=_table_name(),
        IndexName="userIndex",
        KeyConditions={
            "userId": {
                "ComparisonOperator": "EQ",
                "AttributeValueList": [{"S": user_id}],
            }
        },
    )
    return [deposit_from_ddb_item(item) for item in response.get("Items", [])]


def get_deposit_by_id(deposit_id: str) -> Deposit:
    if not deposit_id:
        raise ValueError("Deposit not found")
    response = _ddb().get_item(
        TableName=_table_name(),
        Key={"depositId": {"S": deposit_id}},
        ConsistentRead=True,
    )
    item = response.get("Item")
    if not item:
        raise ValueError("Deposit not found")
    return deposit_from_ddb_item(item)
